File: 51-n-queens/51-n-queens.py
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        ans = []
        board = []
        s = ""
        leftRow = []
        upDia = []
        lowDia = []
        for i in range(n):
            s +="."
            
        for i in range(n):
            board.append(s)
        
        for i in range(n):
            leftRow.append(0)
            
        for i in range(2*n):
            upDia.append(0)
        
        for i in range(2*n):
            lowDia.append(0)
        
        
        def rec(col, board):
            if col==n:
                ans.append(board[:])
                return
            
            for row in range(n):
                if(leftRow[row]==0 and lowDia[col+row]==0 and upDia[n-1+col-row]==0):
                    leftRow[row]=1
                    lowDia[col+row]=1 
                    upDia[n-1+col-row]=1
                    board[row] = board[row][:col] + "Q" + board[row][col+1:]
                                      
                    rec(col+1, board)
                    
                    leftRow[row]=0
                    lowDia[col+row]=0 
                    upDia[n-1+col-row]=0                    
                    board[row] = board[row][:col] + "." + board[row][col+1:]
        
        
        rec(0, board)
        
        return ans

File: 51-n-queens/51-n-queens.js
/**
 * @param {number} n
 * @return {string[][]}
 */
var solveNQueens = function(n) {
    let s=Array(n+1).join(".")
    let board=Array(n).fill(s)
    let ans=[]
    
    let left=Array(n).fill(0)
    let upD=Array(2*n).fill(0)
    let lowD=Array(2*n).fill(0)
    
    const helper = (col, board) => {
        if(col==n){
            ans.push(board.slice())
            return
        }
        for(let row=0; row<n; row++){
            let sArr;
            if(left[row]==0 && upD[(n-1)+(col-row)]==0 && lowD[row+col]==0){
                left[row] = 1 
                upD[(n-1)+(col-row)] = 1 
                lowD[row+col] = 1                
                sArr = board[row].split("")
                sArr[col]="Q"
                board[row] = sArr.join("")
                
                helper(col+1, board)
                
                left[row] = 0
                upD[(n-1)+(col-row)] = 0
                lowD[row+col] = 0
                sArr = board[row].split("")
                sArr[col]="."
                board[row] = sArr.join("")
            }
        }
    }
    
    helper(0, board)
    return ans
};